File: src/slices/player.ts
import * as v from 'valibot';
import type { StateCreator } from 'zustand';

import { getLocalStorageValue } from '../_core/localStorage';
import type { StoreState } from '../store/create';
import type { TimeRange } from '../storeFx/playerContext';

export const REPLAY_GAIN_LOCAL_STORAGE_KEY = 'replayGain';

export enum PreferredGain {
  Album = 'album',
  Track = 'track',
}

const ReplayGainOptions = v.object({
  preAmp: v.number(),
  preferredGain: v.optional(v.enum(PreferredGain)),
});

export type ReplayGainOptions = v.InferInput<typeof ReplayGainOptions>;

export interface PlayerSlice {
  buffered: TimeRange[];
  currentSongId?: string;
  currentTime: number;
  duration: number | undefined;
  muted: boolean;
  paused: boolean;
  queuedSongIds: string[];
  replayGainOptions: ReplayGainOptions;
  volume: number;
}

export const playerSlice: StateCreator<
  StoreState,
  [],
  [],
  PlayerSlice
> = (): PlayerSlice => {
  return {
    buffered: [],
    currentTime: 0,
    duration: undefined,
    muted: false,
    paused: true,
    queuedSongIds: [],
    replayGainOptions: getLocalStorageValue(
      REPLAY_GAIN_LOCAL_STORAGE_KEY,
      ReplayGainOptions,
    ).getOr(() => ({ preAmp: 0 })),
    volume: 1,
  };
};

File: src/store/player.ml
open State

let replay_gain_local_storage_key = "replayGain"

let make () =
  let open Dom.Storage in
  let replayGainOptions =
    localStorage
    |> getItem replay_gain_local_storage_key
    |> Option.map (Fun.compose ReplayGainOptions.of_json Json.parseOrRaise)
    |> Option.value
         ~default:ReplayGainOptions.{ preAmp = 0.0; preferredGain = None }
  in
  {
    buffered = [||];
    currentSongId = None;
    currentTime = 0.0;
    duration = None;
    muted = false;
    paused = true;
    queuedSongIds = [||];
    replayGainOptions;
    volume = 1.0;
  }
